src/robot/ros/RosRobot.cpp

```diff
@@ -38,6 +52,262 @@ RosRobot::RosRobot(
   }
 }
 
+//==============================================================================
+void RosRobot::step(const std::chrono::system_clock::time_point& timepoint)
+{
+  // Lock only if root robot
+  std::unique_ptr<std::lock_guard<std::mutex>> lock;
+  if (!mParentRobot)
+  {
+    lock = std::make_unique<std::lock_guard<std::mutex>>(
+        getRootSkeleton()->getMutex());
+  }
+
+  if(mRosJointModeCommandClient)
+    mRosJointModeCommandClient->step();
+
+  if (!mActiveExecutor.empty())
+  {
+    mExecutors[mActiveExecutor]->step(timepoint);
+  }
+
+  for (const auto& subrobot : mSubRobots)
+  {
+    subrobot.second->step(timepoint);
+  }
+}
+
+//=============================================================================
+RobotPtr RosRobot::registerSubRobot(
+    dart::dynamics::ReferentialSkeletonPtr refSkeleton, const std::string& name)
+{
+  // Ensure name is unique
+  if (mSubRobots.find(name) != mSubRobots.end())
+  {
+    dtwarn << "Subrobot '" << name << "' already exists." << std::endl;
+    return nullptr;
+  }
+
+  // Ensure all body nodes in skeleton are owned by this robot
+  for (auto bodyNode : refSkeleton->getBodyNodes())
+  {
+    if (!mMetaSkeleton->hasBodyNode(bodyNode))
+    {
+      dtwarn << "Subrobot '" << name << "'' contains body node "
+             << bodyNode->getName() << " not in parent MetaSkeleton."
+             << std::endl;
+      return nullptr;
+    }
+  }
+
+  // Ensure subrobot DoFs are disjoint
+  for (const auto& subrobot : mSubRobots)
+  {
+    auto dofs = subrobot.second->getDofs();
+    for (std::string dofName : util::dofNamesFromSkeleton(refSkeleton))
+    {
+      if (dofs.find(dofName) != dofs.end())
+      {
+        dtwarn << "Subrobot '" << name << "'' overlaps existing subrobot "
+               << subrobot.first << " at DoF " << dofName << "." << std::endl;
+        return nullptr;
+      }
+    }
+  }
+
+  // Create the sub RosRobot.
+  auto subRosRobot = std::make_shared<RosRobot>(
+      refSkeleton, this, mCollisionDetector, mSelfCollisionFilter, name);
+  mSubRobots[name] = subRosRobot;
+  return subRosRobot;
+}
+
+//=============================================================================
+void RosRobot::deactivateExecutor()
+{
+  if (!mActiveExecutor.empty())
+  {
+    if (!mRosControllerNames[mActiveExecutor].empty()
+        && !stopController(mRosControllerNames[mActiveExecutor]))
+    {
+      throw std::runtime_error(
```

Comment:
Verify that this does not throw if the controller is already stopped.

Ideally, this would only throw if the controller is started, but cannot shutdown (which is a safety issue). If the controller is already stopped, this should just ignore the error.